modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileTransferServiceImpl.java

```diff
@@ -390,75 +414,64 @@ public CompletableFuture<List<Path>> download(String sourceNodeConsistentId, Ide
-                    })
-                    .whenComplete((v, e) -> {
-                        if (e != null) {
-                            LOG.error("Failed to download files. Identifier: {}", identifier, e);
-                        }
-
-                        IgniteUtils.deleteIfExists(directory);
-                    });
-
-            return messagingService.invoke(sourceNodeConsistentId, FILE_TRANSFER_CHANNEL, downloadRequest, responseTimeout)
-                    .thenApply(FileDownloadResponse.class::cast)
-                    .whenCompleteAsync((response, e) -> {
-                        if (e != null) {
-                            fileReceiver.cancelTransfer(transferId, e);
-                        } else if (response.error() != null) {
-                            fileReceiver.cancelTransfer(transferId, buildException(response.error()));
-                        } else {
-                            fileReceiver.receiveFileHeaders(transferId, response.headers());
-                        }
-                    }, executorService)
-                    .thenCompose(v -> downloadedFiles);
-        } catch (Exception e) {
-            return failedFuture(e);
-        }
+        CompletableFuture<List<Path>> downloadedFiles = new CompletableFuture<List<Path>>()
+                .whenComplete((v, e) -> transferIdToConsumer.remove(transferId));
+
+        transferIdToConsumer.put(transferId, new DownloadRequestConsumer(downloadedFiles, targetDir));
+
+        messagingService.invoke(sourceNodeConsistentId, FILE_TRANSFER_CHANNEL, downloadRequest, responseTimeout)
+                .thenApply(FileDownloadResponse.class::cast)
+                .whenComplete((response, e) -> {
+                    if (e != null) {
+                        downloadedFiles.completeExceptionally(e);
+                    } else if (response.error() != null) {
+                        downloadedFiles.completeExceptionally(toException(response.error()));
+                    }
+                });
+
+        return downloadedFiles;
     }
 
     @Override
     public CompletableFuture<Void> upload(String targetNodeConsistentId, Identifier identifier) {
         return getFileProvider(identifier).files(identifier)
-                .thenCompose(files -> {
-                    FileUploadRequest message = messageFactory.fileUploadRequest()
-                            .identifier(identifier)
-                            .headers(fromPaths(messageFactory, files))
-                            .build();
+                .thenCompose(files -> transferFilesToNode(targetNodeConsistentId, UUID.randomUUID(), identifier, files));
+    }
+
+    private CompletableFuture<Void> transferFilesToNode(
+            String targetNodeConsistentId,
+            UUID transferId,
+            Identifier identifier,
+            List<Path> paths
+    ) {
+        if (paths.isEmpty()) {
+            return failedFuture(new FileTransferException("No files to upload"));
```

Comment:
Why is this a problem? What happens if user tries to do this, will he have an error in his log? It's not what he would expect

Reply:
What error will a user see in this case? If it's an error message in a server log, then it's a problem. If it's an exception message in his application, then I'm fine with it